cmd/customd/client/client.go

```diff
@@ -0,0 +1,448 @@
+	out := make(chan BroadcastTxResponse, 1)
+	defer close(out)
+	go cc.BroadcastTxAsync(tx, out)
+	res := <-out
+	return res
+}
+
+func (cc *CustomClient) BroadcastTxSync(tx weave.Tx, timeout time.Duration) BroadcastTxResponse {
+	data, err := tx.Marshal()
+	if err != nil {
+		return BroadcastTxResponse{Error: err}
+	}
+
+	res, err := cc.conn.BroadcastTxSync(data)
+	if err != nil {
+		return BroadcastTxResponse{Error: err}
+	}
+	if res.Code != 0 {
+		return BroadcastTxResponse{Error: errors.WithMessage(fmt.Errorf("CheckTx failed with code %d", res.Code), res.Log)}
+	}
+
+	// and wait for confirmation
+	evt, err := cc.WaitForTxEvent(data, tmtypes.EventTx, timeout)
+	if err != nil {
+		return BroadcastTxResponse{Error: err}
+	}
+
+	txe, ok := evt.(tmtypes.EventDataTx)
+	if !ok {
+		if err != nil {
+			return BroadcastTxResponse{Error: fmt.Errorf("WaitForOneEvent did not return an EventDataTx object")}
+		}
+	}
+
+	return BroadcastTxResponse{
+		Response: &ctypes.ResultBroadcastTxCommit{
+			DeliverTx: txe.Result,
+			Height:    txe.Height,
+			Hash:      txe.Tx.Hash(),
+		},
+	}
+}
+
+func (cc *CustomClient) WaitForTxEvent(tx tmtypes.Tx, evtTyp string, timeout time.Duration) (tmtypes.TMEventData, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	query := tmtypes.EventQueryTxFor(tx)
+
+	uuid := hex.EncodeToString(append(tx.Hash(), cmn.RandBytes(2)...))
+	evts, err := cc.conn.Subscribe(ctx, uuid, query.String())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to subscribe")
+	}
+
+	// make sure to unregister after the test is over
+	defer cc.conn.UnsubscribeAll(ctx, uuid)
+
+	select {
+	case evt := <-evts:
+		return evt.Data.(tmtypes.TMEventData), nil
+	case <-ctx.Done():
+		return nil, errors.New("timed out waiting for event")
+	}
+}
+
+// BroadcastTxAsync can be run in a goroutine and will output
+// the result or error to the given channel.
+// Useful if you want to send many tx in parallel
+func (cc *CustomClient) BroadcastTxAsync(tx weave.Tx, out chan<- BroadcastTxResponse) {
+	data, err := tx.Marshal()
+	if err != nil {
+		out <- BroadcastTxResponse{Error: err}
+		return
+	}
+
+	// TODO: make this async, maybe adjust return value
+	res, err := cc.conn.BroadcastTxCommit(data)
+	msg := BroadcastTxResponse{
+		Error:    err,
+		Response: res,
+	}
+	out <- msg
+}
+
+// SubscribeHeaders queries for headers and starts a goroutine
+// to typecase the events into Headers. Returns a cancel
+// function. If you don't want the automatic goroutine, use
+// Subscribe(QueryNewBlockHeader, out)
+func (cc *CustomClient) SubscribeHeaders(out chan<- *Header) (func(), error) {
+	query := QueryNewBlockHeader
+	pipe, cancel, err := cc.Subscribe(query)
+	if err != nil {
+		return nil, err
+	}
+	go func() {
+		for msg := range pipe {
+			evt, ok := msg.Data.(tmtypes.EventDataNewBlockHeader)
+			if !ok {
+				// TODO: something else?
+				panic("Unexpected event type")
```

Comment:
Both TODO and panic should be cleaned up :nail_care:

Reply:
What would be the right way to handle this situation?